File: 2026/09/24/bench.py
import time, sys

N = 10_000_000
buf = [None] * 1024
parts = [chr(97 + k) for k in range(16)]

def loop_only(n):
    b = buf
    for i in range(n):
        b[i & 1023] = None

def literal(n):
    b = buf
    for i in range(n):
        b[i & 1023] = ""

def ctor(n):
    b = buf
    for i in range(n):
        b[i & 1023] = str()

def from_int(n):
    b = buf
    for i in range(n):
        b[i & 1023] = str(i)

def concat(n):
    b = buf; p = parts
    for i in range(n):
        b[i & 1023] = p[i & 15] + "x"

cases = [("loop only", loop_only), ('""', literal), ("str()", ctor),
         ("str(i)", from_int), ("p[i&15] + 'x'", concat)]

print(f"Python {sys.version.split()[0]}")
for name, f in cases:
    best = float("inf")
    for _ in range(5):
        t0 = time.perf_counter_ns()
        f(N)
        best = min(best, time.perf_counter_ns() - t0)
    ns = best / N
    print(f"{name:16s} {ns:7.2f} ns/string  {1e3/ns:8.1f} M/s")

File: 2026/09/24/bench.js
const N = 100_000_000;
const buf = new Array(1024).fill(null);
const parts = Array.from({ length: 16 }, (_, k) => String.fromCharCode(97 + k));

function loop_only(n) { for (let i = 0; i < n; i++) buf[i & 1023] = null; }
function literal(n) { for (let i = 0; i < n; i++) buf[i & 1023] = ""; }
function ctor(n) { for (let i = 0; i < n; i++) buf[i & 1023] = String(); }
function from_int(n) { for (let i = 0; i < n; i++) buf[i & 1023] = String(i); }
function concat(n) { for (let i = 0; i < n; i++) buf[i & 1023] = parts[i & 15] + "x"; }

const cases = [["loop only", loop_only], ['""', literal], ["String()", ctor],
  ["String(i)", from_int], ["p[i&15] + 'x'", concat]];

const rt = typeof Bun !== "undefined" ? `Bun ${Bun.version}` : `Node ${process.version}`;
console.log(rt);
for (const [name, f] of cases) {
  f(1_000_000); // warm up
  let best = Infinity;
  for (let r = 0; r < 5; r++) {
    const t0 = process.hrtime.bigint();
    f(N);
    const dt = Number(process.hrtime.bigint() - t0);
    if (dt < best) best = dt;
  }
  const ns = best / N;
  console.log(`${name.padEnd(16)} ${ns.toFixed(2).padStart(7)} ns/string  ${(1e3 / ns).toFixed(1).padStart(8)} M/s`);
}
